node/pkg/watchers/evm/finalizers/optimism.go

```diff
@@ -0,0 +1,142 @@
+		finalizerMapping: make([]FinalizerEntry, 0),
+	}
+}
+
+func (f *OptimismFinalizer) IsBlockFinalized(ctx context.Context, block *connectors.NewBlock) (bool, error) {
+	finalizedL1Block := f.l1Finalizer.GetLatestFinalizedBlockNumber()
+	if finalizedL1Block == 0 {
+		// This happens on start up.
+		return false, nil
+	}
+
+	type Result struct {
+		Mode       string
+		EthContext struct {
+			BlockNumber uint64 `json:"blockNumber"`
+			TimeStamp   uint64 `json:"timestamp"`
+		} `json:"ethContext"`
+		RollupContext struct {
+			Index         uint64 `json:"index"`
+			VerifiedIndex uint64 `json:"verifiedIndex"`
+		} `json:"rollupContext"`
+	}
+
+	var m Result
+	err := f.connector.RawCallContext(ctx, &m, "rollup_getInfo")
+	if err != nil {
+		f.logger.Error("failed to get rollup info", zap.String("eth_network", f.connector.NetworkName()), zap.Error(err))
+		return false, err
+	}
+	if m.RollupContext.VerifiedIndex == 0 {
+		return false, fmt.Errorf("Received a verified index of 0.  Please check Optimism RPC parameter.")
+	}
+
+	f.logger.Debug("finalizerMapping", zap.Uint64("L2 verified index", m.RollupContext.VerifiedIndex), zap.String(" => ", ""), zap.Uint64("L1_blockNumber", m.EthContext.BlockNumber))
+	// Look at the last element of the array and see if we need to add this entry
+	if len(f.finalizerMapping) != 0 && f.finalizerMapping[len(f.finalizerMapping)-1].l2Block > m.RollupContext.VerifiedIndex {
+		return false, fmt.Errorf("The received verified index just went backwards. Received %d. Last number in array is %d", m.RollupContext.VerifiedIndex, f.finalizerMapping[len(f.finalizerMapping)-1].l2Block)
+	}
+	if len(f.finalizerMapping) == 0 || f.finalizerMapping[len(f.finalizerMapping)-1].l2Block < m.RollupContext.VerifiedIndex {
+		// New information.  Append it to the array.
+		f.finalizerMapping = append(f.finalizerMapping, FinalizerEntry{l2Block: m.RollupContext.VerifiedIndex, l1Block: m.EthContext.BlockNumber})
+		f.logger.Info("Appending new entry.", zap.Int("finalizerMap size", len(f.finalizerMapping)), zap.Uint64("L2 verified index", m.RollupContext.VerifiedIndex), zap.Uint64("L1_blockNumber", m.EthContext.BlockNumber))
+	}
+
+	// In order to be finalized, need to:
+	// - walk the finalizerMapping
+	// - Find the first entry <= the L2 block of interest
+	// - Grab the L1 block for that L2 block
+	// - Check to see if that L1 block was finalized
+	// - if so, return true, else return false
+
+	isFinalized := false
+	myBlockNumber := block.Number.Uint64()
+	// Walk the array
+	pruneIdx := -1
+	for idx, entry := range f.finalizerMapping {
+		if entry.l2Block >= myBlockNumber {
+			if finalizedL1Block >= entry.l1Block {
+				isFinalized = true
+				f.logger.Info("found in finalizerMapping", zap.Uint64("L2", myBlockNumber))
+			}
+			// Only prune up to the L2 passed in, which may be different than what we looked at in the array
+			if entry.l2Block == myBlockNumber {
+				pruneIdx = idx
+			}
+			break
+		} else {
+			pruneIdx = idx
```

Comment:
The correctness of this function should not depend on implementation details in some other part of the program (especially when those assumptions are not documented in the code).  So whatever is happening in the rest of guardiand is not really relevant here.  See also: [action at a distance](https://en.wikipedia.org/wiki/Action_at_a_distance_(computer_programming))

Also, your example shows the case where everything goes right but my point is that there still exist scenarios in which the current implementation will return the incorrect result.

* We call IsBlockFinalized for block N, rollup_getInfo() returns N with ETH block M.  ETH block M is finalized so we return true and also, because of lines 96-99 we will prune the returned info:
```go
			// Only prune up to the L2 passed in, which may be different than what we looked at in the array
			if entry.l2Block == myBlockNumber {
				pruneIdx = idx
			}
```
* Now we call IsBlockFinalized for block N-100 and this time rollup_getInfo() returns N+10 with ETH block M+10.  If ETH block M+10 is not finalized, this function will incorrectly return false because we only return true in the case that `finalizedL1Block >= entry.l1Block` (lines 92-95).


Basically, it seems like there is an assumption in the pruning code that IsBlockFinalized will never be called for older blocks.  I am just pointing out that this is not the case and that we still need to store *some* information from the pruned entries.

Reply:
This watcher code is ONLY used by the poller.  The poller is guaranteed to only move forward.  It will never move backward.  This code is NEVER used for re-observations.  But even if the poller had a hiccup where a lower number block was used, this code would still work and that block would eventually be identified as finalized.  I believe that based on this context, this code is reasonable and correct.  This code is not meant to be used as a general purpose library function.

Reply:
Everything you said is true today but is not actually documented anywhere.  Someone might start using the PollFinalizer interface in the future for something else and may not realize that the correctness of this function in a completely different part of the codebase depends on the input parameter monotonically increasing.  This kind of undocumented inter-dependence is very hard to debug when things go wrong.

I don't think it's unreasonable to ask that this function should work correctly independently of the input parameters.  The code change is not that large and I even provided a code snippet that does most of the rework already.

Reply:
I agree we should design individual components to work indepentently of being implicitly coupled to any other component.

Reply:
Added code to keep the latestFinalizedL2Block